### ISSUE TITLE
Importing transactions via FinTS/HBCI

### ISSUE THREAD

**Issue by moay** [firefly-iii/firefly-iii#1159]
I would like to implement a importer for FinTS/HBCI. There is a PHP package available that I'd like to use: https://github.com/mschindler83/fints-hbci-php

As far as I can see, there is a pretty straightforward importing process, so I'd need to create an importer with prerequisites, a configurator and an importing routine. Are there any drawbacks or special stuff I should know about before beginning this feature?

I'd like to provide a way to store HBCI logins and account mappings. Would you prefer storing this in config files or in a database table?

**Comment by JC5** [comment]
Depending on how many logins the user could have, I would suggest storing them in a Preference object. ``Preference::set / get``. Otherwise, use config files (JSON).

* Add the necessary lines to `config/import.php`. I suggest auto-start = true.
* The prerequisites routine should be fairly straightforward. Don't use it for job or bank specific stuff, but for FinTS things only.
* The Spectre import has a redirect / back to Firefly III thing. 
* Use configuration["stage"] to know where your job is at.

I would following the lead of the file / spectre importer. For consistency, use the following config vars

```
'import-account'        => 0, // none,
'apply-rules'           => true,
'match-bills'           => false,
'auto-start'            => false,
'has-config-file'       => true,
```

These can be picked up by other pages.

**Comment by JC5** [comment]
Otherwise, nothing special from the top of my head. If you create a PR and work in the PR I can give pointers where necessary. I appreciate your efforts! :grin:

**Comment by pree** [comment]
I would really like an HBCI connection 👍

**Comment by moay** [comment]
Working on it. It will take a few weeks (at least) as I sadly don't have much time for the project right now. But I definitely need it for my private finances, too, so we'll get there... :-)

**Comment by JC5** [comment]
There's no need to reply to these issues to indicate you want the feature as well guys. We know it :wink:

**Comment by Iretos** [comment]
@moay I also need this feature. Can I help you implement it?

**Comment by moay** [comment]
Hey @Iretos. Actually, I'm kinda stuck in too many things right now. Like I predicted, I don't find the time for side projects whilst growing a six month old and building a house. ;-)

So feel free to implement this - I'll get back to this when I have the time to provide a good solution.

I started by building the base layer of information and assimilated a database of banks (which you can find here: https://github.com/moay/php-fints-database and depend on via composer).

**Comment by Iretos** [comment]
I want to use the FakeRoutine as first example, but when I execute it, I get "Job could not be started or crashed: The import storage routine crashed: Trying to get property of non-object". @JC5 what is wrong with the FakeRoutine?

**Comment by JC5** [comment]
Nothing that I know of. I would not have shipped a broken routine. But bugs can always occur. Could you share the log files?

**Comment by JC5** [comment]
It seems to work on the demo site. If you can share more details I’ll look into it ASAP.

**Comment by Iretos** [comment]
`[2018-06-15 21:50:17] local.ERROR: The import storage routine crashed: Trying to get property of non-object  
[2018-06-15 21:50:17] local.ERROR: #0 /var/www/firefly-iii/app/Http/Controllers/Import/JobStatusController.php(197): FireflyIII\Http\Controllers\Import\JobStatusController->storeTransactions(Object(FireflyIII\Models\ImportJob))
#1 [internal function]: FireflyIII\Http\Controllers\Import\JobStatusController->store(Object(FireflyIII\Models\ImportJob))
#2 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): call_user_func_array(Array, Array)
#3 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): Illuminate\Routing\Controller->callAction('store', Array)
#4 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Route.php(212): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(FireflyIII\Http\Controllers\Import\JobStatusController), 'store')
#5 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Route.php(169): Illuminate\Routing\Route->runController()
#6 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Router.php(659): Illuminate\Routing\Route->run()
#7 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#8 /var/www/firefly-iii/app/Http/Controllers/Import/JobStatusController.php(56): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#9 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(133): FireflyIII\Http\Controllers\Import\JobStatusController->FireflyIII\Http\Controllers\Import\{closure}(Object(Illuminate\Http\Request), Object(Closure))
#10 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#11 /var/www/firefly-iii/app/Http/Controllers/Controller.php(109): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#12 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(133): FireflyIII\Http\Controllers\Controller->FireflyIII\Http\Controllers\{closure}(Object(Illuminate\Http\Request), Object(Closure))
#13 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#14 /var/www/firefly-iii/app/Http/Middleware/Binder.php(77): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#15 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): FireflyIII\Http\Middleware\Binder->handle(Object(Illuminate\Http\Request), Object(Closure))
#16 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#17 /var/www/firefly-iii/app/Http/Middleware/Range.php(63): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#18 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): FireflyIII\Http\Middleware\Range->handle(Object(Illuminate\Http\Request), Object(Closure))
#19 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#20 /var/www/firefly-iii/app/Http/Middleware/AuthenticateTwoFactor.php(78): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#21 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): FireflyIII\Http\Middleware\AuthenticateTwoFactor->handle(Object(Illuminate\Http\Request), Object(Closure))
#22 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#23 /var/www/firefly-iii/app/Http/Middleware/Authenticate.php(73): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#24 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): FireflyIII\Http\Middleware\Authenticate->handle(Object(Illuminate\Http\Request), Object(Closure))
#25 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#26 /var/www/firefly-iii/vendor/laravel/passport/src/Http/Middleware/CreateFreshApiToken.php(50): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#27 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Laravel\Passport\Http\Middleware\CreateFreshApiToken->handle(Object(Illuminate\Http\Request), Object(Closure))
#28 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#29 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(67): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#30 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle(Object(Illuminate\Http\Request), Object(Closure))
#31 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#32 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#33 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\View\Middleware\ShareErrorsFromSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#34 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#35 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(63): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#36 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Session\Middleware\StartSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#37 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#38 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#39 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle(Object(Illuminate\Http\Request), Object(Closure))
#40 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#41 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(59): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#42 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Cookie\Middleware\EncryptCookies->handle(Object(Illuminate\Http\Request), Object(Closure))
#43 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#44 /var/www/firefly-iii/app/Http/Middleware/Sandstorm.php(58): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#45 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): FireflyIII\Http\Middleware\Sandstorm->handle(Object(Illuminate\Http\Request), Object(Closure))
#46 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#47 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#48 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Router.php(661): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#49 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Router.php(636): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#50 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Router.php(602): Illuminate\Routing\Router->runRoute(Object(Illuminate\Http\Request), Object(Illuminate\Routing\Route))
#51 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Router.php(591): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#52 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(176): Illuminate\Routing\Router->dispatch(Object(Illuminate\Http\Request))
#53 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#54 /var/www/firefly-iii/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#55 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Fideloper\Proxy\TrustProxies->handle(Object(Illuminate\Http\Request), Object(Closure))
#56 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#57 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(31): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#58 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle(Object(Illuminate\Http\Request), Object(Closure))
#59 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#60 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(31): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#61 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle(Object(Illuminate\Http\Request), Object(Closure))
#62 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#63 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#64 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle(Object(Illuminate\Http\Request), Object(Closure))
#65 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#66 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(51): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#67 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(151): Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(Object(Illuminate\Http\Request), Object(Closure))
#68 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#69 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#70 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(151): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#71 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(116): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#72 /var/www/firefly-iii/public/index.php(76): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#73 {main} `

**Comment by JC5** [comment]
Mmm, I can't reproduce this and I have no idea what exactly is going wrong. Could you [enable debug mode](https://github.com/firefly-iii/help/wiki/Enable-debug-logging-and-debug-mode) and try again?

**Comment by bnw** [comment]
Hi, 
I implemented a simple FinTS importer based on https://github.com/mschindler83/fints-hbci-php at https://github.com/bnw/firefly-iii/tree/develop.

There are still some things that should be done, e.g.
* Storing config files for later use.
* Translation to german (FinTS is mostly used in Germany)
* Tests

Even without these the feature can be already useful and I think it would be a good idea to merge it.
@JC5 what do you think?

**Comment by JC5** [comment]
Looking good. Some comments:

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-a77cd5dbe739222208431f0b2de3c3a4R24

Can't store password in plaintext ever. Use `Crypt::encrypt`and `Crypt::decrypt`. 

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-a77cd5dbe739222208431f0b2de3c3a4R65

This needs type hinting.

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-a77cd5dbe739222208431f0b2de3c3a4R84

These need `use` statements, it's cleaner.

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-af353b6993acc38b84d9cf7ac6209dacR65

Did you double check if the constructor for MessageBag can take an argument like this? I just don't know.

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-a39ba0283f5f10eb7e9f3ade712ef9e9R62

Use `===` here.

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-a39ba0283f5f10eb7e9f3ade712ef9e9R68

Is this enough to get all accounts back from the FinTS provider thing?

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-a39ba0283f5f10eb7e9f3ade712ef9e9R86

Here's a var dump and try to use `""` instead of `''`.

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-28339589b1e804a8a19ef50a78899091R59

I saw it in more places. I strongly recommend Carbon\Carbon over \DateTime. Unless the class demands it.

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-28339589b1e804a8a19ef50a78899091R82

Use `bcmul($amount, '-1');` to prevent floating point operator errors.

https://github.com/bnw/firefly-iii/commit/61506facf0795dc7aa57baf978698c348ece125c#diff-28339589b1e804a8a19ef50a78899091R113

Check out the code for other import routines (you're behind on master :wink:). There's a 'transaction_source' field you should add as well.

---

So as you can see from the level of my complaints, this is all looking pretty well. Do you have any sites or things where I can test this too?

**Comment by bnw** [comment]
Thanks for your feedback!

1. Can't store password in plaintext ever. Use Crypt::encryptand Crypt::decrypt. :+1: done
2. This needs type hinting. - :+1: done
3. These need use statements, it's cleaner. - :+1: done
4. Did you double check if the constructor for MessageBag can take an argument like this? I just don't know. - :+1: fixed
5. Use === here. - :+1: done
6. Is this enough to get all accounts back from the FinTS provider thing? - ?
7. Here's a var dump and try to use '' instead of "". - :+1: done
8. I saw it in more places. I strongly recommend Carbon\Carbon over \DateTime. Unless the class demands it. - :+1: done
9. Use bcmul($amount, '-1'); to prevent floating point operator errors. - :+1: done
10. Check out the code for other import routines (you're behind on master 😉). There's a 'transaction_source' field you should add as well. - :+1: done
11. Do you have any sites or things where I can test this too? - Unfortunately, I could not find any public "test banks" or API sandboxes. So I guess the only way for you to test it would be to create an account at a german bank that supports it :/

**Comment by JC5** [comment]
Regarding number 6 and the structure of your code. It seems from the $config you’re building you can only import into one Firefly III account at a time. Is that a limit of FinTS or is it something that can be expanded upon? For example, if I have a savings and a checking account. Do I have to run the import routine twice?

**Comment by JC5** [comment]
That is something the other imports do "semi automated". They match on IBANs. I assume that's what you mean with "SEPA-number"?

Looking at the code and the features, I would suggest some improvements in the way of:

- The ability to import multiple accounts in one go.
- The option to apply Firefly III rules.
- The option to either: automatch IBANs or show a screen for selection. I can build this as well if necessary.

Looking at the "testability" of this code: if you can provide (anonymised) example outputs I can play around with those.

But if you create a PR for version 0.1 we can continue already.

**Comment by bnw** [comment]
> I assume that's what you mean with "SEPA-number"?   

Yes! Sorry for the confusion.

---

>Looking at the "testability" of this code: if you can provide (anonymised) example outputs I can play around with those.

Unfortunately the raw FinTS-Messages are super complex. Looking at the raw messages I could not really figure out what is what and don't feel like I can savely anonymize them. Also, there are dozens of messages exchanged just to list a few transactions.
If you want to manually test the the import as a whole I created a mock for the `FinTS` class. If you include the file https://pastebin.com/raw/mjRAEtiX somewhere globaly (e.g. at the end of `bootstrap/app.php`) the FinTS import should accept any credentials and will show two accounts with some transactions.
I think I will also create some automated tests based on this.

---

I believe all your suggestions above should be implemented.
I might implement some automated tests this weekend. If you want implement any of the other features, go ahead :)

I will create a MR now.

**Comment by JC5** [comment]
For that, you can create a method in either tests/TestCase.php or create a trait in tests somewhere. A global include is not the best idea.

**Comment by JC5** [comment]
I've just pushed a commit (see above my reply) in which I've refactored some of your code. Most of it was suggested by my IDE, some things I did myself. There should be no big functional changes. I've also pushed the language files to Crowdin for translation.

**Comment by pree** [comment]
Thank you @bnw for this awesome work so far !
I have a couple of german bank accounts that support FinTS and would like to help you testing. I will upgrade my firefly-iii this week (probably on the weekend) and test the first implementation.

**Comment by pree** [comment]
I've just tried to import transactions via Fints in the current develop docker container. 
It didn't work and threw a couple of errors into the logs: https://gist.github.com/pree/45348e5c854b5ac1272130e349eb41ab

![image](https://user-images.githubusercontent.com/9799276/46916653-b52a0a80-cfbd-11e8-8483-3301375586d5.png)

**Comment by JC5** [comment]
That's an easy fix. I've pushed something, see my commit above this reply.

**Comment by pree** [comment]
Nice! That worked. I was able to import transactions via FinTS now. The only thing thats off that I've noticed is, that the description field isn't parsed, just copied.
HBCI has a a couple of extra informations in those fields which is documented (in german) here on site 207 & 208: https://www.hbci-zka.de/dokumente/spezifikation_deutsch/fintsv3/FinTS_3.0_Messages_Finanzdatenformate_2010-08-06_final_version.pdf 

Basically on SEPA Transactions its 'SVWZ+<Description>'.

In the transactions that I imported I only got SVWZ and EREF (End-to-End Reference) but there are a couple more.

I can try to implement this maybe this weekend or if someone else wants to implement it and doesn't understand the fields in the document I can help with translations

**Comment by JC5** [comment]
For me, "parsing"means that there is content in the "description"-string that has no meaning until an extra translation is made. An example would be the following string:

`Transaction on %date`.

Where `%date` must be "parsed:  into `2018-10-18`.

Is this the case for the FinTS description field as well? Does it contain text that requires expansion, computation or replacement? Or can the description field be expanded into _new_ fields?

If either one is the case (expanding the string or expanding into new fields) this can be done [right here](https://github.com/firefly-iii/firefly-iii/blob/develop/app/Support/Import/Routine/FinTS/StageImportDataHandler.php#L114) where you can see how the transaction is created. The description is filled by simply taking the `$transaction->getDescription1()` field and storing it.

If you wish to expand or split or "parse" the content of this field, this is where you should do it.

As you can see the `$storeData` has a lot of fields. Between `'bunq_payment_id'` and `'original-source'` you can add [these extra meta fields](https://github.com/firefly-iii/firefly-iii/blob/develop/app/Factory/TransactionJournalFactory.php#L118).

More info about the SEPA fields can be found [right here](https://github.com/firefly-iii/firefly-iii/blob/develop/resources/lang/en_US/import.php#L308).

**Comment by pree** [comment]
Basically there are these different variants of transactions:

`EREF+<end-to-end reference>` A reference unique to the transaction
`KREF+<customer reference>` Not sure, isn't set in any of my transaction, probably a customer id
`MREF+<mandate reference>` The reference of the SEPA-mandate (Same reference on all transactions regarding this mendate)
`CRED+<creditor Identifier>` Seems to be the IBAN of the opposing party
`DEBT+<Originators Identification Code>` Not sure how this looks like, isn't set when `CRED` is set.
`SVWZ+<SEPA transaction description>` The description of the transaction.
`ABWA+<different customer>` If the transaction pays for a 3rd party. Example: If you order something online for someone else who does not have an account and you pay it. Your friend is put into `ABWA`.

And all of those are saved in the FinTS Fields 20-29. Currently they are just appended all together in the Description field, so it looks something like: `EREF+<...>MREF+<...>CRED+<...>SVWZ+<...>` making it not really readable.

Looking into the extra meta fields I would save the values as following:

`EREF` -> `sepa-ct-id`
`KREF` -> Not sure
`MREF` -> `sepa-db`
`CRED` -> `sepa-ct-op` or `sepa-ci`
`DEBT` -> Not sure
`SVWZ` -> `description`
`ABWA` -> Not sure

What do you think about this ?

**Comment by bnw** [comment]
@pree Thanks a lot for testing!

Unfortunately, the FinTS library used here does not seem to provide any tools to parse this field.
I have added a method that at least attempts to extract the `SVWZ` field.
I don't really use the other fields so I only added parsing for the `SVWZ` field for now. 

The data provided by the `$transaction->getDescription1()` does not really seem to be consistent. Often times there are fields missing, sometimes even the `SVWZ` field. In this case, the added method falls back to the entire "unparsed" description.

**Comment by JC5** [comment]
I saw the PR, thanks.

Do you see a possibility to lift this away from a private function, into a helper class perhaps? I would suggest a class called `FireflyIII/Support/FinTS/MetadataParser` that has several public (not static!) methods:

* parseDescription
* parseEndToEndRef()
* Etc.

That makes the functionality to parse these FinTS strings testable (and I would like to see test cases) and exportable (for other projects and users).

Arguments about what the method should return (or not) can be settled and demonstrated in test cases, similar to the [Amount parser test cases](https://github.com/firefly-iii/firefly-iii/blob/master/tests/Unit/Import/Converter/AmountTest.php#L47).

**Comment by bnw** [comment]
@JC5 I added a `FireflyIII/Support/FinTS/MetadataParser` and some tests for it in the PR https://github.com/firefly-iii/firefly-iii/pull/1797

**Comment by JC5** [comment]
I saw that, nice! I have committed some small changes, please check out the commit my reply. There are mainly some IDE suggestions. Thanks! :+1:

**Comment by JC5** [comment]
I'm closing this one as I consider the original question to be solved with the implementation of FinTS. Please open new tickets if problems arise.
